server/lib/security-factory/security-factory.ts

```diff
@@ -1,23 +1,34 @@
 import { OpendistroFactory, XpackFactory, DefaultFactory } from './factories';
 import { KibanaRequest, RequestHandlerContext } from 'src/core/server';
-import { PluginSetup } from '../../types'
+import { PluginSetup } from '../../types';
 
 type CurrentUser = {
-  username?: string 
-  authContext: {[key:string]: any}
-}
+  username?: string;
+  authContext: { [key: string]: any };
+};
 
 export interface ISecurityFactory {
-  platform?: string
-  getCurrentUser(request: KibanaRequest, context?:RequestHandlerContext): Promise<CurrentUser> 
+  platform?: string;
+  getCurrentUser(request: KibanaRequest, context?: RequestHandlerContext): Promise<CurrentUser>;
 }
 
-export function SecurityObj({security, opendistroSecurityKibana}:PluginSetup): ISecurityFactory {
-  if (!!security) {
-    return new XpackFactory(security);
-  } else if (!!opendistroSecurityKibana) {
-    return new OpendistroFactory(opendistroSecurityKibana);
-  } else {
-    return new DefaultFactory();
+export async function SecurityObj(
+  { security, opendistroSecurityKibana }: PluginSetup,
+  context?: RequestHandlerContext
+): Promise<ISecurityFactory> {
+  const params = {
+    path: `/_security/user`,
+    method: 'GET',
+  };
+
+  try {
+    const responseCurl = await context.core.elasticsearch.client.asInternalUser.transport.request(
+      params
+    );
+  } catch (error) {
+    return !!opendistroSecurityKibana
+      ? new OpendistroFactory(opendistroSecurityKibana)
+      : new DefaultFactory();
   }
```

Comment:
**nitpick**: The request to check if X-Pack is enabled should be done only if `security` plugin is installed.